pyo3-derive-backend/src/frompy.rs

```diff
@@ -0,0 +1,428 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::punctuated::Punctuated;
+use syn::token::Paren;
+use syn::{
+    parse_quote, Attribute, DataEnum, DeriveInput, Expr, ExprCall, Fields, Ident, PatTuple, Result,
+    Variant,
+};
+
+/// Describes derivation input of an enum.
+#[derive(Debug)]
+struct Enum<'a> {
+    enum_ident: &'a Ident,
+    vars: Vec<Container<'a>>,
+}
+
+impl<'a> Enum<'a> {
+    /// Construct a new enum representation.
+    ///
+    /// `data_enum` is the `syn` representation of the input enum, `ident` is the
+    /// `Identifier` of the enum.
+    fn new(data_enum: &'a DataEnum, ident: &'a Ident) -> Result<Self> {
+        if data_enum.variants.is_empty() {
+            return Err(syn::Error::new_spanned(
+                &data_enum.variants,
+                "Cannot derive FromPyObject for empty enum.",
+            ));
```

Comment:
Iirc, it depends on the attributes for serde, without adding `tag(type)` or whatever the attribute is, it will serialize empty structs & variants the same as `None`, e.g. as `null` in `serde_json`. I think deserialization works almost the same, it's possible to deserialize `SomeEnum::Foo` from the serialization of `SomeEnum::Bar` if the enum is untagged.

In any way, I think having a `#[transparent]` variant around `&'a PyAny` is a better solution for a catch-all extraction than discarding the input and generating an empty variant. At least, using such an empty variant as catch-all is the only reasonable use-case that I could come up with, unless we want to tie such variants to something like type names.